src/optimization/Combiner.cpp

```diff
@@ -1121,6 +1125,518 @@ InstructionWalker optimizations::combineArithmeticOperations(
+                    auto mpLiteral = mpValue.has_value() ? mpValue->getLiteralValue() : Optional<Literal>{};
+
+                    if(mpLiteral)
+                    {
+                        if(mpLiteral->unsignedInt() < (1u << 12))
+                        {
+                            auto it = bb.walk();
+                            bool firstCall = true;
+                            while(!it.isEndOfBlock())
+                            {
+                                auto call = it.get<intermediate::MethodCall>();
+                                if(call && std::find(loadInstrs.begin(), loadInstrs.end(), call) != loadInstrs.end())
+                                {
+                                    it.erase();
+
+                                    auto output = *call->getOutput();
+                                    if(firstCall)
+                                    {
+                                        firstCall = false;
+
+                                        auto addrArg = call->assertArgument(1);
+
+                                        auto elemType = addrArg.type.getElementType();
+                                        auto vectorSize = elemType.getInMemoryWidth() * vectorLength;
+
+                                        // TODO: limit loadInstrs.size()
+                                        Value offset = assign(it, TYPE_INT32) =
+                                            offsetValues[0] * Literal(vectorLength * elemType.getInMemoryWidth());
+                                        Value addr = assign(it, TYPE_INT32) = offset + addrArg;
+
+                                        uint16_t memoryPitch =
+                                            static_cast<uint16_t>(mpLiteral->unsignedInt()) * vectorSize;
+
+                                        DataType VectorType{
+                                            elemType.getInMemoryWidth() * DataType::BYTE, vectorLength, false};
+
+                                        uint64_t rows = loadInstrs.size();
+                                        auto entries = Value(Literal(static_cast<uint32_t>(rows)), TYPE_INT32);
+                                        it = method.vpm->insertReadRAM(method, it, addr, VectorType, nullptr, true,
+                                            INT_ZERO, entries, Optional<uint16_t>(memoryPitch));
+
+                                        VPMArea area(VPMUsage::SCRATCH, 0, static_cast<uint8_t>(rows));
+                                        it = method.vpm->insertReadVPM(method, it, output, &area, true);
+                                    }
+                                    else
+                                    {
+                                        // TODO: gather these instructions in one mutex lock
+                                        it = method.vpm->insertLockMutex(it, true);
+                                        assign(it, output) = VPM_IO_REGISTER;
```

Comment:
If we use the "scratch" area, then this does not work, since the "scratch" area might be overridden by another QPU. For this to work, we would need to use an own VPM area per QPU.